main/lsp/json_types.cc

```diff
@@ -175,4 +175,31 @@ unique_ptr<Diagnostic> Diagnostic::copy() const {
     return d;
 }
 
+string TextDocumentContentChangeEvent::apply(string oldSource) const {
+    if (range) {
+        auto &r = *range;
+        // incremental update
+        core::Loc::Detail start, end;
+        start.line = r->start->line + 1;
+        start.column = r->start->character + 1;
+        end.line = r->end->line + 1;
+        end.column = r->end->character + 1;
+        core::File old("./fake/path.rb", string(oldSource), core::File::Type::Normal);
+        // These offsets are non-nullopt assuming the input range is a valid range.
+        auto startOffset = core::Loc::pos2Offset(old, start).value();
+        auto endOffset = core::Loc::pos2Offset(old, end).value();
+        return oldSource.replace(startOffset, endOffset - startOffset, text);
+    } else {
+        return text;
+    }
+}
+
+string DidChangeTextDocumentParams::getSource(string_view oldFileContents) const {
+    string rv = string(oldFileContents);
+    for (auto &change : contentChanges) {
+        rv = change->apply(move(rv));
```

Comment:
Pardon my C++ noobness, but once you `move(rv)`, I thought `rv` was invalidated?  Do you need to move line 198 inside the `for` loop?

Reply:
Oh, duh, I'm misreading the line.  I see now!